### ISSUE TITLE
Codebase is coded against base classes classes.

### ISSUE THREAD

**Issue by ndobromirov** [lcache/lcache#64]
### Steps to reproduce
Code review.

### Expected behavior
Every moving piece from the library to be covered by an interface.

### Actual behavior
There are no interfaces and everything is coded against a base class or something similar, kind-of breaking Interface segregation principle (from SOLID).

There should be clear interfaces for L1, L2 and integration layer to make this fully pluggable.
If there are repeated parts from L1/L2/Integration, they should be moved to a separate interface.
This will allow more shallow hierarchies and documentation re-usage.

### Frameworks tested
- [ ] Using lcache/lcache directly

### Proposal
Introduce segregated interfaces for what's needed at the moment.
If there is something to be reused you can always move it to a trait.

Have L1, L2 and integration interfaces that extend the small ones.
Redactor the code-base to utilize the interfaces as small as possible.

### Notes
This is just an idea that popped when reading the code. Feel free to discard if deemed invalid.

**Comment by stevector** [comment]
Yeah, I think it would be sensible to do this work prior to a 1.0.0 release. Doing so will help make it clear which features/decisions can be changed inside 1.x.x and which changes would have to wait until a 2.0.0.

**Comment by davidstrauss** [comment]
The issue with standard approaches to isolation from the base classes is the differing constructor needs for different L1 and L2 implementations. Some implementations need a PDO handle, while others need nothing. If I added a RedisL2, it would require an active Redis handle. So, the normal factory pattern isn't great here because it either becomes loaded with optional parameters or loses native type checking.

We could provide a factory for each implementation (or set of similar implementations), but changing the factory you use depending on what you want defeats half the value of factories.

Unfortunately, PHP lacks the concept of [friend classes](https://en.wikipedia.org/wiki/Friend_class), which would be the proper way to handle this issue.

**Comment by greg-1-anderson** [comment]
I don't think there is anything wrong with:
```
class RedisL2 extends L2
{
     static function create($pdo) { ... }
}

$myL2 = RedisL2::create($pdo);
```
I think that the main point of the suggestion in this issue is that functions that take untyped parameters, such as [Integrated::__construct()](https://github.com/lcache/lcache/blob/master/src/Integrated.php#L11), should instead take typed parameters, e.g. `L1Interface`, `L2Interface`, or maybe just `LXInterface`. The code is making some assumption about what methods are available on these objects; these assumptions should be formalized with type-checked parameters.  Factories are of a secondary concern; you make your cache objects once at the beginning of the program, and then you're done. For this sort of use-case, I think that static factory methods are entirely appropriate.

**Comment by davidstrauss** [comment]
It's probably unrelated to your point, but the reason I raised the concept of a RedisL2 is that it would not take a PDO object; it would take a Redis handle. Even if a Redis handle is an object, it wouldn't share any type hierarchy with PDO. So, type-checking set on the superclass or interface level isn't possible.

Getting back to the point, defining a function like `create()` at the superclass or interface level would require defining at least the parameter count, and that already breaks things like StaticL2 that don't take any parameters for construction.

As long as the static creation functions are on the classes themselves, there's no more isolation than today versus the other public methods on the classes. I don't see how telling developers to only call `$l2 = DatabaseL2::create($pdo)` is better than telling developers to only call `$l2 = new DatabaseL2($pdo)`. In either case, the public methods called by other parts of LCache are accessible to the library user.

What I'm looking for here is a concrete way for how changes of this sort will improve the developer experience or maintenance of this library.

**Comment by davidstrauss** [comment]
As for the typed parameters for constructing LCache\Integrated, I completely agree. It should take an `LCache\L1` and an `LCache\L2`, but I still don't see how using interfaces adds value beyond the abstract superclasses.

The current superclasses aren't heavy; they just contain naive implementations of things like existence checks that may have more optimal implementation for some persistence stores. Forcing use of a superclass, in effect, sets a more rigorous interface by saying, "If you override any of these things, the behavior must be identical to the naive implementation you're overriding."

**Comment by greg-1-anderson** [comment]
`static function create($pdo) ` was a typo: should have been RedisHandle as you said. Sorry--I feel like that one mistake has taken this conversation down the wrong path, as the main point of the OP, I think, it typehinting, not factory methods.

As far as factory methods go, though, the factory methods themselves do not need to conform to any specific interface.  Every static factory method has its own method signature.  The `create` method is not part of the interface. You could put this static method in a separate RedisL2Factory object if that bothers you, but I don't think that's necessary.

The advantage of a static factory function is that, at some point in the future, the implementation of DatabaseL2::create() could change (provided that its parameter list did not) to return some other class of object.  With `new` you are always stuck with that one class. For simple code, just bump semver and break everyone; no big deal. I don't think it is super important here; just use `new` if you prefer.

As for your second point about interfaces over abstract superclasses, sure, that's fine -- but use type hinting in methods that take the abstract superclass as a parameter, then.  If nothing else is done for this issue, that alone would be beneficial & sufficient.

**Comment by greg-1-anderson** [comment]
One reason I prefer to use interfaces rather than abstract methods for type hinting, though, is that the interface does not contain any implementation, so it's easier to read an interface than an abstract class, if you're just interested in the list of public methods and their prototypes.

**Comment by davidstrauss** [comment]
You'll see no pushback from me on type hinting. :-)

**Comment by davidstrauss** [comment]
I've added the type hints now. I will have to keep thinking about the tension between these various library design goals.

**Comment by ndobromirov** [comment]
This is now easier to solve (at least on L1 level) as all the constructors follow the same signature in #73. 
There is a factory class that handles the creation of any L1 concrete.
